### ISSUE TITLE
element to hide in mode:exam

### ISSUE THREAD

**Issue by trombonekenny** [PrairieLearn/PrairieLearn#6365]
Questions sometimes include reference links that need the Internet that do not work in Internet-limited testing centers.

We should modify pl-hide-in-panel or make a similar element to hide content when "mode": "Exam" and encourage instructors to update their questions appropriately.

**Comment by echuber2** [comment]
Here's another aspect to consider:

This would also be useful for questions where instructors want to include extra hints _except_ at exam time. However, that benefit would be lost for students taking remote/async exams that aren't in Exam mode. (I know exam mode is different from exam _type_, but for reference, see @nwalters512's objection here: https://github.com/PrairieLearn/PrairieLearn/discussions/6108#discussioncomment-3321833 ) So if there's a risk that instructors might try to use this for anything other than marking up network-requiring items, it could be an antipattern.

Perhaps an alternative approach could be that individual questions that need external internet access could explicitly (but optionally?) declare this in their `info.json`, and then an Exam mode assessment would throw an error (or at sync time?) if any such questions are included. I'm thinking the declaration in the question's json wouldn't actually toggle network access, it would only assist in this kind of static checking at the assessment level.

**Comment by echuber2** [comment]
To avoid confusion about what such a property in `info.json` actually does, perhaps it could be named something like `examModeCompatible` (default `true`) rather than mentioning "allow internet" or "require internet" in the name. However instructors would be need to be taught to set this to `false` if their question requires external net access, which may also be tough to communicate.

**Comment by trombonekenny** [comment]
My main motivation is to stop unavailable links from showing up in questions in the CBTF. The students think they're getting short changed because they can't reference something in the question, they complain to the proctors, there's troubleshooting about what's wrong, etc. I just want to make that easier for all.

I can't see having faculty having the presence of mind to add more things into info.json, and then have to do things with it. I'm doubtful they would even use a `<pl-hide-in-mode>` tag unless we harass them about it.

It's probably easiest (?) to somehow parse the question at view time and put a warning/pop-up/something on the links that we know are blocked (i.e. not PrairieLearn) so students at least know the missing link is part of the question design not the CBTF misbehaving.

**Comment by trombonekenny** [comment]
This is relevant again in some current discussions about what's visible in exams. Currently `data` used with the elements and external graded doesn't have any information about the environment in which the question is run. I think adding `data["mode"]` would be useful in this context and hopefully not too controversial (as adding other information into the question visibility, like who is taking it or which assessment it's included in, etc.)

Where in the code does the `data` dict that's shipped to the question server.py, element controller python, and external grader data.json live? (I'm hoping one place but I wouldn't be surprised if it's three....)

**Comment by nwalters512** [comment]
@trombonekenny see here for the data to be passed to `render()`:

https://github.com/PrairieLearn/PrairieLearn/blob/7b955833c6802454e7f9121fd69efc31853bcb4f/apps/prairielearn/src/question-servers/freeform.js#L1139

TBD if we need this for `prepare()` and `generate()` too. I'm not convinced we need to provide this information to external graders; do you have a use case in mind?

Rather than `data["mode"]`, consider `data["assessment_mode"]` or `data["assessment_access_mode"]` (bikeshed away). I'd rather namespace this.

Be sure to think through how to define the "mode" when the question is viewed outside the context of an assessment.

**Comment by mwest1066** [comment]
I'm still very skeptical about giving questions access to `mode`. The whole concept of `mode` is something that's already confusing, and having questions behave differently based on this mode scares me. I'm also not sure what the use case is here?

**Comment by nwalters512** [comment]
> My main motivation is to stop unavailable links from showing up in questions in the CBTF. The students think they're getting short changed because they can't reference something in the question, they complain to the proctors, there's troubleshooting about what's wrong, etc. I just want to make that easier for all.
>
> ...
>
> It's probably easiest (?) to somehow parse the question at view time and put a warning/pop-up/something on the links that we know are blocked (i.e. not PrairieLearn) so students at least know the missing link is part of the question design not the CBTF misbehaving.

I think this is legit. But if this is the only use case, I imagine we could do this by intercepting click events client-side and prompting students with a modal explaining access may be blocked if they're in exam mode. I haven't yet thought through exactly how good of an idea that'd be, but it's an option and would require zero intervention from instructors.

**Comment by eliotwrobson** [comment]
Isn't this effectively a more limited version of this existing PR?: https://github.com/PrairieLearn/PrairieLearn/pull/3848
This is effectively wanting questions to behave differently when given different injectable parameters (except here the parameter is just the assessment context and is not set manually). This also couples the mode to the question code itself, and I'm not sure what the behavior should be if the assessment types ever get changed (which is planned at some point in the future, if I remember right).

I'm also in agreement with @mwest1066 that having different static question behavior has the potential to cause issues, and it would be nice to have some feature to aid question development with respect to this.

**Comment by jonatanschroeder** [comment]
@eliotwrobson this is access mode, not assessment type. That said, I think linking this to the assessment via params as Eliot mentioned is probably a better idea. I think it is a bad idea for students to see different question text depending on mode. If the difference is between different assessments it makes sense, but within the assessment the student should probably always see the same question text.

**Comment by trombonekenny** [comment]
Let my XY this a bit. I think we might be able to do it with CSS and change the display based on context (without doing it as an element, which prompted this question).

**Comment by reteps** [comment]
This came up recently with CS128's course, and again with Phill Conrad's course.

**Comment by pconrad** [comment]
Yes, this would be a very useful feature.

**Comment by jonatanschroeder** [comment]
CSS might not be a good idea if the intention is to hide information in exam mode, since students would have access to the information by checking the source or developer mode.

**Comment by mwest1066** [comment]
I agree with @eliotwrobson that https://github.com/PrairieLearn/PrairieLearn/pull/3848 is the right approach to handle this.

**Comment by pconrad** [comment]
But it appears that #3848 and it's successor have been closed without being merged.

Is that approach still on the roadmap, or has it been ruled out?

**Comment by eliotwrobson** [comment]
The successor PR is here https://github.com/PrairieLearn/PrairieLearn/pull/11946

**Comment by reteps** [comment]
This is now possible with question preferences. Closed via #14014. UI coming in #14478

**Comment by pconrad** [comment]
> This is now possible with question preferences. Closed via [#14014](https://github.com/PrairieLearn/PrairieLearn/pull/14014). UI coming in [#14478](https://github.com/PrairieLearn/PrairieLearn/pull/14478)

Can you explain how?  I read through the issues and I'm at a loss.

What I want to be able to do is to check at render time what kind of assessment the question is a part of, and based on that, either render a hint, or not.  Something like:

```
{{#somethingGoesHere}}
<pl-hint>
   A hint appears here, but only when the question is NOT part of an exam.
</pl-hint>  
{{/somethingGoesHere}}}
```

You say: "it's now possible with question preferences", but if it's only possible with secret, undocumented knowledge, it isn't really possible.

**Comment by nwalters512** [comment]
The idea is that instead of using assessment types (which will not exist in the long run, so we didn't build _exactly_ what was asked for), you add preferences (https://docs.prairielearn.com/question/preferences/) to your questions - for instance, a `show_hint` preference. Then, when building your assessment, you toggle that to an appropriate value - e.g. `true` on a homework or `false` on quizzes/exams. You can then read `{{#preferences.show_hint}}` in your `question.html` file, or `data["preferences"]["show_hint"]` in `server.py`. This exact use case is in fact documented as an example: https://docs.prairielearn.com/question/preferences/#adjusting-difficulty

**Comment by pconrad** [comment]
> The idea is that instead of using assessment types (which will not exist in the long run, so we didn't ...

Thanks!  This is helpful!
